apps/api/src/common/authority-checker.service.ts

```diff
@@ -0,0 +1,380 @@
+import { PrismaClient, Authority, Workspace } from '@prisma/client'
+import { VariableWithProjectAndVersion } from 'src/variable/variable.types'
+import {
+  BadRequestException,
+  Injectable,
+  NotFoundException,
+  UnauthorizedException
+} from '@nestjs/common'
+import getCollectiveProjectAuthorities from './get-collective-project-authorities'
+import getCollectiveWorkspaceAuthorities from './get-collective-workspace-authorities'
+import { EnvironmentWithProject } from 'src/environment/environment.types'
+import { ProjectWithSecrets } from 'src/project/project.types'
+import { SecretWithProjectAndVersion } from 'src/secret/secret.types'
+
+export interface AuthorityInput {
+  userId: string
+  entity: {
+    id?: string
+    name?: string
+  }
+  authority: Authority
+  prisma: PrismaClient
+}
+
+@Injectable()
+export class AuthorityCheckerService {
+  public async checkAuthorityOverWorkspace(
+    input: AuthorityInput
+  ): Promise<Workspace> {
+    const { userId, entity, authority, prisma } = input
+
+    let workspace: Workspace
+
+    try {
+      if (entity?.id) {
+        workspace = await prisma.workspace.findUnique({
+          where: {
+            id: entity.id
+          }
+        })
+      } else {
+        workspace = await prisma.workspace.findFirst({
+          where: {
```

Comment:
I think we need to tweak this a bit. Workspaces are unique across user level, but not global level. So essentially this might end up in a case where we fetch a workspace that is not in question at all. 

We can drill into the `members` property. Here, we can check if the provided `userId` is `in` the list or not. That way, we will only get the entities that the user belongs to. If the result is null, we throw an error as usual.